### PR TITLE
Use a `gen` in `PartialEq` for `UseSharedState`

### PR DESCRIPTION
**Pull request by DanielAlbl**
So I had tried using a `UseSharedState` as a dependency for `use_future`, and it wasn't re-running when it changed. I think this was because it was comparing the `.inner` values which still were pointing to the same data. 
I implemented a special `UseFutureDep` so it would compare a `gen` value, sort of following the example of `UseRef`. I did have to remove the `PartialEq` here to avoid conflicting with this: `impl<A: Dep> UseFutureDep for &A`, but this might not be the right thing to do.

### CODE REVIEW

File: packages/hooks/src/use_shared_state.rs
Comment (Demonthos) [inline packages/hooks/src/use_shared_state.rs:312]: Can we make this the partialeq impl instead like use state does?

UseFutureDep is implemented for all tuples with partial eq elements. This may  fix your problem for a single dependency, but it will make use shared state unusable with multiple dependencies in a tuple

**Review by Demonthos — CHANGES_REQUESTED** [review]

File: packages/hooks/src/use_shared_state.rs
Comment (DanielAlbl) [reply packages/hooks/src/use_shared_state.rs]: OK this makes sense. I put a `gen` in `UseSharedState` itself to use in `PartialEq`, since the `gen`s in `ProvidedStateInner` will always be the same. I have them getting synced up in `use_shared_state`, similar to how `use_state` updates `current_val`. I'm not sure if this is the simplest solution, but it does seem to be working in tuples now.

**Review by ealmloff — APPROVED** [review]
Sorry I lost track of this. Looks good, thank you!

**Comment by ealmloff** [comment]
Clippy failure is unrelated. This change can potentially be breaking (this could fail to memorize some props). Waiting for a second review before merging

**Pull request merged**
